src/platform/plugins/shared/workflows_execution_engine/server/step/nodes_factory.ts

```diff
@@ -51,138 +68,133 @@ export class NodesFactory {
     private workflowGraph: WorkflowGraph
   ) {}
 
-  public create<TStep extends BaseStep>(
-    step: TStep // TODO: TStep must refer to a node type, not BaseStep (IfElseNode, ForeachNode, etc.)
-  ): NodeImplementation {
-    const stepType = (step as any).type; // Use a more type-safe way to determine step type if possible
-    const stepId = (step as any).id || (step as any).name;
-    if (!stepType) {
-      throw new Error('Step type is not defined for step: ' + JSON.stringify(step));
-    }
-
+  public create<TStep extends UnionExecutionGraphNode>(node: TStep): NodeImplementation {
     const stepLogger = this.workflowLogger.createStepLogger(
       this.workflowRuntime.getCurrentStepExecutionId(),
-      stepId,
-      stepId,
-      stepType
+      node.stepId,
+      node.stepId,
+      node.stepType
     );
 
-    // Check if it's an internal action type
-    const isElasticsearchAction = stepType && stepType.startsWith('elasticsearch.');
-    const isKibanaAction = stepType && stepType.startsWith('kibana.');
+    // Handle elasticsearch.* and kibana.* actions
+    if (node.stepType && node.stepType.startsWith('elasticsearch.')) {
+      this.workflowLogger.logInfo(`Creating Elasticsearch action step: ${node.stepType}`, {
+        event: { action: 'internal-action-creation', outcome: 'success' },
+        tags: ['step-factory', 'elasticsearch', 'internal-action'],
+      });
+      return new ElasticsearchActionStepImpl(
+        node as any,
+        this.contextManager,
+        this.workflowRuntime,
+        this.workflowLogger
+      );
+    }
+
+    if (node.stepType && node.stepType.startsWith('kibana.')) {
+      this.workflowLogger.logInfo(`Creating Kibana action step: ${node.stepType}`, {
+        event: { action: 'internal-action-creation', outcome: 'success' },
+        tags: ['step-factory', 'kibana', 'internal-action'],
+      });
+      return new KibanaActionStepImpl(
+        node as any,
+        this.contextManager,
+        this.workflowRuntime,
+        this.workflowLogger
+      );
+    }
 
-    switch (stepType) {
+    switch (node.type) {
       case 'enter-foreach':
         return new EnterForeachNodeImpl(
-          step as any,
+          node as EnterForeachNode,
```

Comment:
there's no need for casting here actually, the node type is inferred because of switch

https://github.com/user-attachments/assets/4d59afbe-922b-4332-b48c-76799b7dd067